pkg/config/leaderelection/leaderelection.go

```diff
@@ -29,8 +29,14 @@ import (
 // See https://github.com/kubernetes/kubernetes/issues/107454 for
 // details on how to migrate to "leases" leader election.
 // Don't forget the callbacks!
-// TODO: In the next version we should switch to using "leases"
 func ToLeaderElectionWithConfigmapLease(clientConfig *rest.Config, config configv1.LeaderElection, component, identity string) (leaderelection.LeaderElectionConfig, error) {
+	return ToLeaderElection(clientConfig, config, component, identity, resourcelock.ConfigMapsLeasesResourceLock)
+}
+
+// ToLeaderElection returns a leader election config that you just need to fill in the Callback for. The input
+// resourceLock must be a value from the k8s.io/client-go/tools/leaderelection/resourcelock package.
+// Don't forget the callbacks!
+func ToLeaderElection(clientConfig *rest.Config, config configv1.LeaderElection, component, identity, resourceLock string) (leaderelection.LeaderElectionConfig, error) {
```

Comment:
The Endpoint and ConfigMap locks are not supported by upstream since Kubernetes v1.24 so the only locking mechanisms that should be supported are lease, endpointlease and configmaplease. Since there is hardly a way to limit if any string is allowed to be passed, it would be better to write two functions:
- ToLeaderElectionWithEndpointLease
- ToLeaderElectionWithLease

And make ToLeaderElection internal to prevent people from using unsupported features.

Reply:
I disagree. I don't think this library should limit the lease options based on what is supported or unsupported in the latest Kubernetes version. If the user of the library is running against an older version of Kubernetes (e.g. OpenShift 3.11), they wouldn't be able to use the lease API.

Also, rather than try to keep up with new lease options in client-go as client-go gets updated, I'd rather just rely on its validation here rather than add further restrictions in this library:
https://github.com/kubernetes/client-go/blob/e6998df99e84a57bfe500bf6fa96e9a724b1698c/tools/leaderelection/resourcelock/interface.go#L165-L211

Reply:
@tkashem do you have an opinion on that since you work on the initial move from ConfigMap to ConfigmapLease in https://github.com/openshift/library-go/pull/1282.

Reply:
> I disagree. I don't think this library should limit the lease options based on what is supported or unsupported in the latest Kubernetes version

library-go can be a bit opinionated, in the interest of convergence, ideally we want OpenShift operators to start using lease (i can't see a reason why a user would have opinion to use configmap vs endpoint here, the leader election feature on the surface is invariant no matter what the underlying lock is)


Can you help us understand your actual use case here?

Reply:
@tkashem my main goal is to exclusively use the lease API for leader election, but having support for configmapsleases allows consumers of the library to update to the latest version of the library without making the switch just yet. OpenShift can later decide when to change the default and/or remove the configmapsleases option.

I personally don't need the endpoint lease option, but was asked by @dgrisonnet to add it in a comment above. I'll remove it now that the library attempts to be opinionated.

Reply:
My initial comment was meant to reduce the scope of resources to only the ones that are supported today, but it is fine with me if we just want to add leases now. The things I didn't want was to make it possible to use leader election with configmaps and secrets as these are not supported anymore.